### DISCUSSION TITLE
Redirect 404s to proxy

### DISCUSSION THREAD

**Discussion by crummy** [vikejs/vike#987] [Discussion] [answered]
I'd like the following behaviour in dev mode:

1. Request comes in
2. Attempt to handle it with VPS
3. If no page exists, do not redirect to 404, instead proxy to another server

I know I can sort of do this with config like this:

```
  server: {
    port: 5173,
    proxy: {
      "/services": "http://localhost:4568",
      "/resources": "http://localhost:4568",
      [...]
    },
  },
```
 However the list of endpoints is quite long... this is not a practical way to do it. 

Can I add my own 404 handler to do this with VPS?

**Comment by brillout** [comment, marked as answer]
See conversation at https://github.com/brillout/vite-plugin-ssr/discussions/986. Does that answer your question?

**Reply by crummy** [reply]
I've added logging events in my `renderer` folder to:

* +onBeforeRender.ts
* +onBeforeRoute.ts
* +onPrerenderStart.ts
* +onRenderClient.tsx
* +onRenderHtml.tsx

The only one I saw when viewing a 404 page was `onBeforeRoute`. But I don't know if it's a 404 in `onBeforeRoute`'s handler so not sure how to handle it there. In #986 it's suggested I can catch them in other hooks but at least judging by the missing logging statements they aren't being reached?

**Reply by brillout** [reply]
That's because you don't have a `/pages/_error/+Page.js`. In that case you can simply check whether `pageContext.httpResponse === null`.

**Reply by crummy** [reply]
Got it, thanks.

I was running my app with plain `vite` but by hosting the vite middleware with a node server I was able to add special handling for 404s with http-proxy.
